.github/workflows/test-go.yml

```diff
@@ -52,79 +52,109 @@ on:
+        uses: hashicorp/vault-action@130d1f5f4fe645bb6c83e4225c04d64cfb62de6e
+        with:
+          url: ${{ steps.vault-auth.outputs.addr }}
+          caCertificate: ${{ steps.vault-auth.outputs.ca_certificate }}
+          token: ${{ steps.vault-auth.outputs.token }}
+          secrets: |
+            kv/data/github/${{ github.repository }}/datadog-ci DATADOG_API_KEY;
+            kv/data/github/${{ github.repository }}/github-token username-and-token | github-token;
+            kv/data/github/${{ github.repository }}/license license_1 | VAULT_LICENSE_CI;
+            kv/data/github/${{ github.repository }}/license license_2 | VAULT_LICENSE_2;
+            kv/data/github/${{ github.repository }}/hcp-link HCP_API_ADDRESS;
+            kv/data/github/${{ github.repository }}/hcp-link HCP_AUTH_URL;
+            kv/data/github/${{ github.repository }}/hcp-link HCP_CLIENT_ID;
+            kv/data/github/${{ github.repository }}/hcp-link HCP_CLIENT_SECRET;
+            kv/data/github/${{ github.repository }}/hcp-link HCP_RESOURCE_ID;
+      - id: setup-git-private
+        name: Setup Git configuration (private)
+        if: github.repository == 'hashicorp/vault-enterprise'
+        run: |
+          git config --global url."https://${{ steps.secrets.outputs.github-token }}@github.com".insteadOf https://github.com
+      - id: setup-git-public
+        name: Setup Git configuration (public)
+        if: github.repository != 'hashicorp/vault-enterprise'
+        run: |
+          git config --global url."https://${{ secrets.ELEVATED_GITHUB_TOKEN}}@github.com".insteadOf https://github.com
+      - run: go install gotest.tools/gotestsum@v1.9.0
+
+      - run: mkdir -p test-results/go-test
+
+      # We use a unique "read-" prefix to guarantee that we're not scribbling on
+      # the aggregated test data in the event of test failure. This key is
+      # unique for every test run and just used to restore the previous
+      # aggregated data. We persist all test data after a successful run and
+      # store that in the go-test-reports- cache.
+      - id: restore-from-cache
+        uses: actions/cache@88522ab9f39a2ea568f7027eddc7d8d8bc9d59c8 # v3.3.1
+        with:
+          path: test-results/go-test
+          key: read-go-test-reports-${{ github.run_number }}
+          restore-keys: go-test-reports-
+      - name: List cached results
+        id: list-cached-results
+        run: ls -lhR test-results/go-test
+      - name: Build matrix excluding binary tests
+        id: build-non-binary
+        env:
+          GOPRIVATE: github.com/hashicorp/*
+        run: |
+          (
+            go list ./... | grep -v "_binary" | gotestsum tool ci-matrix --debug \
+              --partitions 16 \
+              --timing-files 'test-results/go-test/*.json' > matrix.json
+          )
+      - name: Capture list of binary tests
+        id: list-binary-tests
+        run: |
+          LIST="$(go list ./... | grep "_binary" | xargs)"
```

Comment:
We should probably just do this in its own step and store the results for re-use.